[ENH] dilation mapping transformer

fspinna opened the pull request:
I am trying to implement a simple transformer for applying dilation mapping to time series data, as seen in:
```bib
@article{schafer2023weasel,
        title={WEASEL 2.0--A Random Dilated Dictionary Transform for Fast,
        Accurate and Memory Constrained Time Series Classification},
        author={Sch{\"a}fer, Patrick and Leser, Ulf},
        journal={arXiv preprint arXiv:2301.10194},
        year={2023}
   }
```
Basically, it just reorders the element (and indexes) of a time series to imitate a dilation for models that do not support it.

I am currently failing five tests running check_estimator

```
FAILED: test_fit_transform_output[DilationMappingTransformer-TransformerFitTransformSeriesUnivariate]
FAILED: test_fit_transform_output[DilationMappingTransformer-TransformerFitTransformSeriesMultivariate]
FAILED: test_fit_transform_output[DilationMappingTransformer-TransformerFitTransformSeriesUnivariateWithY]
FAILED: test_fit_transform_output[DilationMappingTransformer-TransformerFitTransformPanelUnivariateWithClassY]
FAILED: test_fit_transform_output[DilationMappingTransformer-TransformerFitTransformPanelUnivariateNumpyWithClassYOnlyFit]
```
The error seems to be this one 

```
TypeError: scitype must be a string or list of strings
```

The function that raises the error is 

```
scitype = _coerce_list_of_str(scitype, var_name="scitype")
````
because scitype=None

I am wondering if I am doing something wrong with the tags maybe?
    


<!--
##### For all contributions
- [ ] I've added myself to the [list of contributors](https://github.com/sktime/sktime/blob/main/CONTRIBUTORS.md) with any new badges I've earned :-)
  How to: add yourself to the [all-contributors file](https://github.com/sktime/sktime/blob/main/.all-contributorsrc) in the `sktime` root directory (not the `CONTRIBUTORS.md`). Common badges: `code` - fixing a bug, or adding code logic. `doc` - writing or improving documentation or docstrings. `bug` - reporting or diagnosing a bug (get this plus `code` if you also fixed the bug in the PR).`maintenance` - CI, test framework, release.
  See here for [full badge reference](https://allcontributors.org/docs/en/emoji-key)
- [ ] Optionally, I've added myself and possibly others to the [CODEOWNERS](https://github.com/sktime/sktime/blob/main/CODEOWNERS) file - do this if you want to become the owner or maintainer of an estimator you added.
  See here for further details on the [algorithm maintainer role](https://www.sktime.net/en/latest/get_involved/governance.html#algorithm-maintainers).
- [ ] The PR title starts with either [ENH], [MNT], [DOC], or [BUG]. [BUG] - bugfix, [MNT] - CI, test framework, [ENH] - adding or improving code, [DOC] - writing or improving documentation or docstrings.

##### For new estimators
- [ ] I've added the estimator to the API reference - in `docs/source/api_reference/taskname.rst`, follow the pattern.
- [ ] I've added one or more illustrative usage examples to the docstring, in a pydocstyle compliant `Examples` section.
- [ ] If the estimator relies on a soft dependency, I've set the `python_dependencies` tag and ensured
  dependency isolation, see the [estimator dependencies guide](https://www.sktime.net/en/latest/developer_guide/dependencies.html#adding-a-soft-dependency).
-->

<!--
Thanks for contributing!
-->

fkiraly commented:
hm, let's see.

My suspicion is, the tags should be 

```python
        "scitype:transform-input": "Series",
        "scitype:transform-output": "Series",
```

since you're also saying the transform should be per-instance (i.e., series).

fspinna commented:
Ok, thanks, I think I've got it. I changed the tags to 'Series', and also modified the mtype to 'pd.DataFrame' instead of 'nested_univ' (which, to my understanding, is not supported for single Series). I thought this was the only way to represent multivariate time series with signals of different lengths, but perhaps this is not yet supported? I also had to reset the index, given that it needs to be monotonic.

##### For all contributions
- [ ] I've added myself to the [list of contributors](https://github.com/sktime/sktime/blob/main/CONTRIBUTORS.md) with any new badges I've earned :-)
  How to: add yourself to the [all-contributors file](https://github.com/sktime/sktime/blob/main/.all-contributorsrc) in the `sktime` root directory (not the `CONTRIBUTORS.md`). Common badges: `code` - fixing a bug, or adding code logic. `doc` - writing or improving documentation or docstrings. `bug` - reporting or diagnosing a bug (get this plus `code` if you also fixed the bug in the PR).`maintenance` - CI, test framework, release.
  See here for [full badge reference](https://allcontributors.org/docs/en/emoji-key)
- [x] The PR title starts with either [ENH], [MNT], [DOC], or [BUG]. [BUG] - bugfix, [MNT] - CI, test framework, [ENH] - adding or improving code, [DOC] - writing or improving documentation or docstrings.

##### For new estimators
- [x] I've added the estimator to the API reference - in `docs/source/api_reference/taskname.rst`, follow the pattern.
- [x] I've added one or more illustrative usage examples to the docstring, in a pydocstyle compliant `Examples` section.

fkiraly commented:
> I thought this was the only way to represent multivariate time series with signals of different lengths

That's not quite correct, we have moved to multi-indexed `pandas` format as nesting of `pandas` is increasingly buggy since the 1.5 and 2.0 versions, also see here:
https://github.com/sktime/sktime/blob/main/examples/AA_datatypes_and_datasets.ipynb

There's also automatic broadcasting across instances of different lengths if you implement the transformation for a single series.

Implementing it for multiple series is only necessary if:

* you want to implement a more efficient option than the default broadcast
* the fitting takes multiple series into account, which I understand your transformation does not?

fspinna commented:
Ok, thank you for the correction. So I think the easiest thing is to implement it for a single series then. 
And if I understood correctly, "univariate-only" should be True, and I guess "capability:unequal_length" is not really relevant for element-wise transforms with mtype "pd.Series"?
I think now it should work as intended.

fkiraly commented:
> And if I understood correctly, "univariate-only" should be True

Only if you want the tarnsformer to work only with univariate series. Application to multivariate series is broadcast over variables then.

> and I guess "capability:unequal_length" is not really relevant for element-wise transforms with mtype "pd.Series"

Yes, for transformations that operate per series, and turn one single series into another series, that one is automatically True. You can set it as true or leave it, in which case it will get the default value (True).

fkiraly commented on docs/source/api_reference/transformations.rst:
```diff
@@ -292,6 +292,14 @@ Depending on the transformer, the transformation parameters can be fitted.
+    :toctree: auto_generated/
+    :template: class.rst
+
+    DilationMappingTransformer
```
This is the wrong section, as the transform is not element-wise, it applies to entire series.

I can't spot a section to put it in, this is a resampling-like operation.
Do you have a good idea?

Perhaps we can change "Binning and segmentation" to "Binning, sampling, and segmentation"? Or, a new sampling section?

fspinna replied:
Yes, you are right, I didn't really know where to put it. It is more like a permutation, but you can also view it as a particular type of sampling. I am thinking if there are other permutation-like things for time series, but the only thing I can think of is maybe for data augmentation, i.e., permuting close timesteps to add a little bit of noise. I don't know what's best for this case. Maybe use "Binning, sampling, and segmentation" for now and change it later if new approaches arise?

fkiraly replied:
sure. Augmentation though usually creates multiple series from one, this transformer creates one series from one, so it doesnt fully fit.

fkiraly reviewed (changes requested):
Looks good!

Small but blocking comment, it's in a section in the API reference where it does not fit well, though I'm not sure where to sort it elsewhere. Some thoughts above.

fkiraly commented on sktime/transformations/series/dilation_mapping.py:
```diff
@@ -0,0 +1,117 @@
+class DilationMappingTransformer(BaseTransformer):
+    r"""DilationMapping transformer.
+
+    A transformer for applying dilation mapping to time series data, as seen in:
```
not blocking, but it would be very useful for the user to explain what it does, mathematically - after all, it is not very complicated. Perhaps just one line or two of math? "dilation" is not a standing term.

fkiraly commented on docs/source/api_reference/transformations.rst:
```diff
@@ -410,7 +410,7 @@ Differencing, slope, kinematics
 
-Binning and segmentation
+Binning, sampling and segmentation
 ~~~~~~~~~~~~~~~~~~~~~~~~
```
you also need to lengthen the underline

fkiraly reviewed (approved):
Thanks.

I updated the docstring a little, kindly have a look whether this is fine:

* moved the reference to the `References` section, so it is properly rendered
* Changed the math explanation so it is a bit clearer
